src/nupic/types/Sdr.cpp

```diff
@@ -0,0 +1,710 @@
+     * order to propigate any changes to the dense & sparse formats.
+     */
+    void SparseDistributedRepresentation::setFlatSparseInplace() {
+        // Check data is valid.
+        #ifdef NTA_ASSERTIONS_ON
+            NTA_ASSERT(flatSparse.size() <= size);
+            for(auto idx : flatSparse) {
+                NTA_ASSERT(idx < size);
+            }
+        #endif
+        // Set the valid flags.
+        clear();
+        flatSparse_valid = true;
+        do_callbacks();
+    }
+
+    /**
+     * Update the SDR to reflect the value currently inside of the sparse
+     * vector. Use this method after modifying the sparse vector inplace, in
+     * order to propigate any changes to the dense & flatSparse formats.
+     */
+    void SparseDistributedRepresentation::setSparseInplace() {
+        // Check data is valid.
+        #ifdef NTA_ASSERTIONS_ON
+            NTA_ASSERT(sparse.size() == dimensions.size());
+            for(UInt dim = 0; dim < dimensions.size(); dim++) {
+                const auto coord_vec = sparse[dim];
+                NTA_ASSERT(coord_vec.size() <= size);
+                NTA_ASSERT(coord_vec.size() == sparse[0].size()); // All coordinate vectors have same size.
+                for(auto idx : coord_vec) {
+                    NTA_ASSERT(idx < dimensions[dim]);
+                }
+            }
+        #endif
+        // Set the valid flags.
+        clear();
+        sparse_valid = true;
+        do_callbacks();
+    }
+
+    /**
+     * Destroy this SDR.  Makes SDR unusable, should error or clearly fail if
+     * used.  Also sends notification to all watchers via destroyCallbacks.
+     * This is a separate method from ~SDR so that SDRs can be destroyed long
+     * before they're deallocated; SDR Proxy does this.
+     */
+    void SparseDistributedRepresentation::deconstruct() {
+        clear();
+        size_ = 0;
+        dimensions_.clear();
+        for( auto func : destroyCallbacks ) {
+            if( func != nullptr )
+                func();
+        }
+    }
+
+    /**
+     * Use this method only in conjuction with sdr.initialize() or sdr.load().
+     */
+    SparseDistributedRepresentation::SparseDistributedRepresentation() {}
+
+    /**
+     * Create an SDR object.  The initial value is all zeros.
+     *
+     * @param dimensions A list of dimension sizes, defining the shape of the
+     * SDR.  The product of the dimensions must be greater than zero.
+     */
+    SparseDistributedRepresentation::SparseDistributedRepresentation( const vector<UInt> dimensions )
+        { initialize( dimensions ); }
+
+    void SparseDistributedRepresentation::initialize( const vector<UInt> dimensions ) {
+        dimensions_ = dimensions;
+        NTA_CHECK( dimensions.size() > 0 ) << "SDR has no dimensions!";
+        // Calculate the SDR's size.
+        size_ = 1;
+        for(UInt dim : dimensions)
+            size_ *= dim;
+	NTA_CHECK(size_ > 0) << "SDR: all dimensions must be > 0";
```

Comment:
enforces what the comment says

Reply:
No, NetworkAPI requires that an empty SDR can be created with a dimension of 0 as a placeholder.

Reply:
> No, NetworkAPI requires that an empty SDR can be created with a dimension of 0 as a placeholder. 

I can add that. Also, should be a test.

Reply:
what we use now is `SDR({0})` which is imho a SDR with dim 1, size 0. 
Would you like to use `SDR({})` (empty set {}) for the placeholder? Which I find ,more true to the meaning. @dkeeney